File: Logistic Regression/Algorithms in C++/logisticRegression.cpp
#include <iostream>
#include <cmath>
#include <vector>

// Define the Logistic Regression class
class LogisticRegression {
private:
    std::vector<double> weights;
    double learningRate;
    int numIterations;

public:
    LogisticRegression(double learningRate, int numIterations)
        : learningRate(learningRate), numIterations(numIterations) {}

    // Sigmoid function
    double sigmoid(double z) {
        return 1 / (1 + exp(-z));
    }

    // Training function
    void train(const std::vector<std::vector<double>>& X, const std::vector<int>& y) {
        int numSamples = X.size();
        int numFeatures = X[0].size();

        // Initialize weights with zeros
        weights.resize(numFeatures, 0.0);

        // Gradient descent optimization
        for (int iter = 0; iter < numIterations; ++iter) {
            std::vector<double> gradients(numFeatures, 0.0);

            // Compute gradients
            for (int i = 0; i < numSamples; ++i) {
                double prediction = predict(X[i]);
                double error = prediction - y[i];

                for (int j = 0; j < numFeatures; ++j) {
                    gradients[j] += error * X[i][j];
                }
            }

            // Update weights
            for (int j = 0; j < numFeatures; ++j) {
                weights[j] -= learningRate * gradients[j];
            }
        }
    }

    // Prediction function
    double predict(const std::vector<double>& sample) {
        double z = 0.0;
        int numFeatures = sample.size();

        for (int j = 0; j < numFeatures; ++j) {
            z += weights[j] * sample[j];
        }

        return sigmoid(z);
    }
};

int main() {
    // Training data
    std::vector<std::vector<double>> X = { {2.0, 1.0}, {4.0, 5.0}, {6.0, 3.0}, {8.0, 7.0} };
    std::vector<int> y = { 0, 1, 0, 1 };

    // Create and train the logistic regression model
    LogisticRegression model(0.1, 1000);
    model.train(X, y);

    // Test data
    std::vector<double> sample = { 3.0, 2.0 };

    // Make a prediction
    double prediction = model.predict(sample);
    std::cout << "Prediction: " << prediction << std::endl;

    return 0;
}

File: Logistic Regression/Algorithms in C#/logisticRegression.cs
using System;

class LogisticRegression
{
    private double[] weights;

    public void Train(double[][] inputs, int[] labels, double learningRate, int numIterations)
    {
        int numFeatures = inputs[0].Length;
        weights = new double[numFeatures + 1]; // +1 for the bias term

        for (int iteration = 0; iteration < numIterations; iteration++)
        {
            for (int i = 0; i < inputs.Length; i++)
            {
                double[] inputWithBias = new double[numFeatures + 1];
                Array.Copy(inputs[i], inputWithBias, numFeatures); // Copy input features
                inputWithBias[numFeatures] = 1.0; // Bias term

                double prediction = Predict(inputWithBias);
                double error = labels[i] - prediction;

                for (int j = 0; j < weights.Length; j++)
                {
                    weights[j] += learningRate * error * inputWithBias[j];
                }
            }
        }
    }

    public double Predict(double[] input)
    {
        double linearOutput = 0.0;

        for (int i = 0; i < input.Length; i++)
        {
            linearOutput += weights[i] * input[i];
        }

        return Sigmoid(linearOutput);
    }

    private double Sigmoid(double x)
    {
        return 1.0 / (1.0 + Math.Exp(-x));
    }
}

class Program
{
    static void Main()
    {
        // Training data
        double[][] inputs = new double[][]
        {
            new double[] { 1.0, 1.0 },
            new double[] { 2.0, 2.0 },
            new double[] { 3.0, 3.0 },
            new double[] { 4.0, 4.0 },
            new double[] { 5.0, 5.0 }
        };

        int[] labels = new int[] { 0, 0, 0, 1, 1 };

        // Create and train logistic regression model
        LogisticRegression logisticRegression = new LogisticRegression();
        logisticRegression.Train(inputs, labels, 0.1, 1000);

        // Test the model
        double[] testInput = new double[] { 6.0, 6.0 };
        double prediction = logisticRegression.Predict(testInput);

        Console.WriteLine("Prediction: " + prediction);
    }
}
